### PR TITLE
SAI proposal for icmp echo offload

### PR DESCRIPTION
**Pull request by chikkaiah-work**
This specification suggests:

    The idea of an ICMP echo object or session that can be transferred to HW/ASIC or other modules for faster link detection.
    The introduction of SAI APIs to outline ICMP echo session properties.
    The introduction of a switch-level notification API to monitor changes in ICMP echo session states.
    The switchover flow on link down detection. 
    
Creating this PR to recommit #1943.

### CODE REVIEW

**Review by zjswhhh — COMMENTED** [review]
Please update with: 
1. Jai's questions in the offline emails
2. LAG solution

**Comment by chikkaiah-work** [comment]
> Please update with:
> 
>     1. Jai's questions in the offline emails
> 
>     2. LAG solution

**Comment by chikkaiah-work** [comment]
closed by mistake

**Comment by chikkaiah-work** [comment]
> > Please update with:
> > ```
> > 1. Jai's questions in the offline emails
> > 
> > 2. LAG solution
> > ```

Added SAI_ICMP_ECHO_SESSION_ATTR_RX_PORT to specify RX port. 
Added doc section "“ICMP echo session with LAG”

File: inc/saiicmpecho.h
Comment (JaiOCP) [inline inc/saiicmpecho.h:83]: I believe hw lookup means that ICMP packet is forwarded using L3 lookup instead of forwarding to a port. Please rename the attribute and add comments to that intent.

File: inc/saiicmpecho.h
Comment (JaiOCP) [inline inc/saiicmpecho.h:103]: This means that session is created between peering links. How do we specify more sessions i.e. monitoring between multiple peering links

File: inc/saiicmpecho.h
Comment (JaiOCP) [inline inc/saiicmpecho.h:215]: This is very fuzzy. What happens if there are multiple such protection NHGs. How does this attribute indicates which one to of the protection NHG takes effect. We just can't trigger all the protection NHG to take over.

File: inc/saiicmpecho.h
Comment (chikkaiah-work) [reply inc/saiicmpecho.h:83]: You are correct. It's intended for L3 lookup for forwarding. I am willing to rename it, but the same naming convention is also used in saitwamp.h and saibfd.h, and I prefer to maintain consistency.  Please let me know if still want me to rename it.

Comment (JaiOCP) [reply inc/saiicmpecho.h:83]: In that case please add details in comments clearly stating what happens when flag is set to true or false

File: inc/saiicmpecho.h
Comment (JaiOCP) [reply inc/saiicmpecho.h:103]: This means that SONiC has to create multiple echo sessions; one per peering ports.
If thats the requirement then you may want to add the CRM attribute in switch. Something like
SAI_SWITCH_ATTR_AVAILABLE_ICMP_ECHO_SESSIONS read only attribute.
Initial read will give the max and as more sessions as added subsequent reads will give remaining.

File: inc/saiicmpecho.h
Comment (JaiOCP) [reply inc/saiicmpecho.h:215]: @chikkaiah-work Please look at this comment as well

Comment (chikkaiah-work) [reply inc/saiicmpecho.h:215]: You brought up a valid point. If it's possible for various NHGs to share the same monitor object, it seems reasonable to consider modifying this attribute to the NHG object ID. Should this attribute be left unset (which is null as default), automatic switchover will not occur. Instead, the switchover should be initiated based on the specified NHG. Please let me know if this sounds fine to you

File: doc/SAI-Proposal-ICMP-ECHO.md
Comment (lolyu) [inline doc/SAI-Proposal-ICMP-ECHO.md:138]: should be `SAI_ICMP_ECHO_SESSION_ATTR_HW_LOOKUP_VALID`

File: doc/SAI-Proposal-ICMP-ECHO.md
Comment (lolyu) [inline doc/SAI-Proposal-ICMP-ECHO.md:99]: should we use the same cookie for sessions on the same device or we can use multiple cookies?

File: doc/SAI-Proposal-ICMP-ECHO.md
Comment (lolyu) [inline doc/SAI-Proposal-ICMP-ECHO.md:252]: Please provide examples with LAG

File: inc/saiicmpecho.h
Comment (lolyu) [reply inc/saiicmpecho.h:215]: if there are new NHGs added, will the new NHGs be modified based on the monitor session?

File: doc/SAI-Proposal-ICMP-ECHO.md
Comment (lolyu) [inline doc/SAI-Proposal-ICMP-ECHO.md:252]: also, please provide examples of sessions with protection group.

Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md:246]: Added example with section tiltled " ICMP echo session with protection group"

Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md:246]: Added a section "ICMP echo session with LAG" and explained session creation process.

File: inc/saiicmpecho.h
Comment (chikkaiah-work) [reply inc/saiicmpecho.h:104]: Added SAI_SWITCH_ATTR_AVAILABLE_ICMP_ECHO_SESSION and SAI_SWITCH_ATTR_MAX_ICMP_ECHO_SESSION.

File: inc/saiicmpecho.h
Comment (chikkaiah-work) [reply inc/saiicmpecho.h:84]: Added a brief description for this attribute.  Also, this attribute is mentioned explained in the examples.

File: inc/saiicmpecho.h
Comment (chikkaiah-work) [reply inc/saiicmpecho.h:216]: @lolyu, you are right. Also I have updated the FRR document by mentioning echo session could be included as monitor object.  I have added an example under the section "ICMP echo session with protection group"

File: doc/SAI-Proposal-ICMP-ECHO.md
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md]: fixed.

File: doc/SAI-Proposal-ICMP-ECHO.md
Comment (chikkaiah-work) [reply doc/SAI-Proposal-ICMP-ECHO.md:99]: Each session can share the same cookie, but if a device needs to communicate with multiple devices, each session can have its own unique cookie. The decision to assign cookies to sessions is implementation-specific, and the device can implement a check to enforce the desired behavior.

File: inc/saiicmpecho.h
Comment (JaiOCP) [reply inc/saiicmpecho.h:84]: Thank You

File: inc/saiicmpecho.h
Comment (JaiOCP) [reply inc/saiicmpecho.h:104]: Thank You

**Review by zjswhhh — APPROVED** [review]
LGTM. 

Please wait for Jai's approval before merging.

File: inc/saiicmpecho.h
Comment (JaiOCP) [reply inc/saiicmpecho.h:216]: Thanks for updating the spec with an example workflow. Looks good to me.

**Review by JaiOCP — APPROVED** [review]

**Comment by chikkaiah-work** [comment]
@kcudnik - can you please help merge this.

**Review by lolyu — APPROVED** [review]

**Review by kcudnik — APPROVED** [review]

**Comment by chikkaiah-work** [comment]
> @kcudnik - can you please help merge this.

@rlhui - can we get this merged please ?

**Pull request merged**
